clients/drcachesim/tests/drcachesim_unit_tests.cpp

```diff
@@ -306,9 +308,201 @@ LLC {
+                std::cerr << "ERROR in unit_test_cache_size():"
+                          << " cache_size=" << cache_size
+                          << " test_buffer_size=" << buffer_size
+                          << " num_accesses=" << read_count << " hits=" << c_stats.hits
+                          << " expected_hits=" << expected_hits
+                          << " misses=" << c_stats.misses
+                          << " expected_misses=" << expected_misses << "\n";
+            }
+            assert(c_stats.hits == expected_hits);
+            assert(c_stats.misses == expected_misses);
+        }
+    }
+}
+
+// Tests a cache to verify its line size works as expected.
+void
+unit_test_cache_line_size()
+{
+    // Range of line sizes to test.
+    constexpr int MIN_LINE_SIZE = 16;
+    constexpr int MAX_LINE_SIZE = 256;
+
+    constexpr int BLOCKS_PER_WAY = 16;
+    constexpr int ASSOCIATIVITY = 2;
+
+    for (int line_size = MIN_LINE_SIZE; line_size <= MAX_LINE_SIZE; line_size *= 2) {
+        // Stride through the cache at a test line size.  If test line size is
+        // less than actual line size, there will be cache hits.  If test line
+        // size is larger than actual line size, there will be fewer misses than
+        // lines in the cache.
+        for (uint32_t stride = line_size / 2; stride < line_size * 2; stride *= 2) {
+            int cache_line_count = BLOCKS_PER_WAY * ASSOCIATIVITY;
+            int total_cache_size = line_size * cache_line_count;
+            cache_t cache;
+            caching_device_stats_t stats(/*miss_file=*/"", line_size);
+            cache.init(ASSOCIATIVITY, line_size, total_cache_size, /*parent=*/nullptr,
+                       &stats);
+            auto read_count =
+                generate_1D_accesses(cache, 0, stride, total_cache_size / stride);
+            CStats c_stats = get_cache_stats(stats);
+
+            int expected_misses =
+                (stride <= line_size) ? cache_line_count : total_cache_size / stride;
+            int expected_hits = read_count - expected_misses;
+            assert(read_count == total_cache_size / stride);
+
+            if (expected_hits != c_stats.hits || expected_misses != c_stats.misses) {
+                std::cerr << "ERROR in unit_test_cache_line_size():"
+                          << " line_size=" << line_size << " stride=" << stride
+                          << " read_count=" << read_count << " hits=" << c_stats.hits
+                          << " expected_hits=" << expected_hits
+                          << " misses=" << c_stats.misses
+                          << " expected_misses=" << expected_misses << "\n";
+            }
+            assert(c_stats.hits == expected_hits);
+            assert(c_stats.misses == expected_misses);
+        }
+    }
+}
+
 int
 main(int argc, const char *argv[])
 {
+    unit_test_cache_line_size();
+    unit_test_cache_size();
+    unit_test_cache_associativity();
```

Comment:
Hmm, seems nicer to have this order match the function definition order above?

Reply:
Done.